In [1]:
from sklearn.datasets import fetch_california_housing
X,y = fetch_california_housing(as_frame=True, return_X_y= True)

## WEEK3 PRACTICE

Instruction: Based on california housing dataset answer the Questions that follow.

Que 1
What is the shape of feature matrix for the dataset?

a. (20640, 8)

b. (2064, 8)

c. (20610, 8)

d. (20600, 8)

Correct Answer: a

In [2]:
X.shape

(20640, 8)

Que 2
How many missing values are present in the dataset?

a. 100

b. 157

c. 0

d. 87

Correct answer: c

In [3]:
X.isna().sum().sum()

0

Instructions: For Que 3 to 6, split the california housing dataset into train and test sets.

[Hint: us appropriate API with random_state=0, shuffle=False and test_size=0.2]

Que 3
What is the value of house age in third sample of test set?

a) 23

b) 28

c) 37

d) 17

Correct Answer: b

Que 4
What is the value of Population in the first sample of training set?

a) 2401

b) 1700

c) 322

d) 157

Correct Anwswer: c

Que 5
What is the mean of output label in training set?

a) 1.98

b) 1.96

c) 2.04

d) 2.02

Correct Answer:d

Que 6
What is the median of output label in test set?

a) 2.02

b) 1.96

c) 2.14

d) 2.34

Correct Answer: c

In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, shuffle=False, random_state=42)

print(X_test.HouseAge.iloc[2])
print(X_train.Population.iloc[0])
print(y_train.mean())
print(y_test.median())

28.0
322.0
2.02067031310562
2.1475


Que 7
If we use StandardScaler for preprocessing and LinearRegression for fitting the model with training set obtained in previous questions, what are the coefficients obtained in the predicted model?

Note: Use methods and objects with default parameters.

a. [ 0.7296 , 0.2187, -0.3655, 0.4056, -0.0045 , -0.0493, -0.9998, -0.8705 ]

b. [ 0.9296 , 0.2287, -0.3655, 0.4056, -0.0045 , -0.0393, -0.8998, -0.8705 ]

c. [ 0.2961, 0.1875, -0.6552, 0.30569623, -0.0045 , -0.0393, -0.8998, -0.8705 ]

d. [ 0.8381, 0.1203, -0.3213, 0.3645, -0.0017,-0.0452, -0.8799, -0.8263]

Correct Answer : d

In [5]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
pipe = Pipeline(steps=[('scaler', StandardScaler()),
                       ('regressor', LinearRegression())])

pipe.fit(X_train, y_train)
pipe[-1].coef_

array([ 0.83817967,  0.12032096, -0.32135834,  0.36456599, -0.00170674,
       -0.04524631, -0.87994217, -0.82634126])

Que 8
If we use StandardScaler for preprocessing and LinearRegression for fitting the model with training set used in previous question, what is the bias term predicted by the model?

Note: Use methods and objects with default parameters

a. 1.06855

b. 2.02067

c. 3.06855

d. -2.06855

Correct Answer : b

In [6]:
pipe[-1].intercept_

2.020670313105614

## Week 3 Graded

Instructions: For all graded questions, split the california housing dataset into train and test sets.

[Hint: use appropriate API with random_state=0, shuffle=False and test_size=0.2]

Que 9
If we use StandardScaler for preprocessing and LinearRegression for fitting the model with training set used in previous question, what is the R squared value for predictions obtained using test set features?

Note: Use methods and objects with default parameters

a. 0.66051

b. 0.70623

c. 0.80623

d. 0.90623

Correct Answer : a

In [7]:
import sklearn.metrics as skm
y_pred = pipe.predict(X_test)
skm.r2_score(y_test, y_pred)

0.6605140591531993

Que 10
If we use StandardScaler for preprocessing and LinearRegression for fitting the model, what is the root mean squared error value for predictions obtained using test set features?

Note: Use methods and objects with default parameters

a. 0.8241

b. 0.9241

c. 0.7033

d. 0.6241

Correct Answer : c

In [8]:
import sklearn.metrics as skm
y_pred = pipe.predict(X_test)
skm.mean_squared_error(y_test, y_pred, squared=False)

0.7033383507521878

Instructions for next ques from Que 11 to 16

Split the data into training and test sets with random_state=0, shuffle=False and test_size=0.2 parameters.

Let, y_test= target label in test set of california housing dataset

and y_pred=target labels obtained by the model using X_test

Then compute values of the following evaluation metrics

Que 11
Explained Variance score

a) 0.66

b) 0.81

c) 0.73

d) 0.90

Correct Answer: a

Que 12
Max Error

a. 0.18

b. 2.16

c. 0.66

d. 7.26

Correct Answer: d

Que 13
Mean Absolute Error

a) 0.213

b) 0.516

c) 0.398

d) 0.422

Correct Answer: b

Que 14
Mean Squared Error

a) 0.038

b) 0.494

c) 0.506

d) 0.872

Correct Answer: b

In [9]:
import sklearn.metrics as skm
def many_scores_at_once(y_true, y_pred):
  Functions = [skm.explained_variance_score, skm.max_error, skm.mean_absolute_error, skm.mean_squared_error, (lambda x,y: skm.mean_squared_error(x,y, squared=False)), skm.r2_score]
  Names = ("Explained variance score:", "Max Error: ", "Mean Absolute Error: ", "Mean Squared Error:", "Root Mean Squared Error: ", "R squared score:")
  Scores = [f(y_true, y_pred) for f in Functions]
  output = dict(zip(Names, Scores))
  return output

many_scores_at_once(y_test, y_pred)


{'Explained variance score:': 0.6605500501742703,
 'Max Error: ': 7.260453292958372,
 'Mean Absolute Error: ': 0.5168526993787042,
 'Mean Squared Error:': 0.4946848356388075,
 'Root Mean Squared Error: ': 0.7033383507521878,
 'R squared score:': 0.6605140591531993}

Instructions: for Ques 15 to 16, perform SGD Regression on the given (scaled using StandardScalar()) dataset, using default hyperparameters and random_state=0.

Que 15
What is the bias term?

a) 2.0112

b) 1.0704

c) 0.0704

d) 3.0704

Correct Answer: a

Que 16
What are the coefficients in predicted model?

a) [ 0.9268, 2.1200, -0.2552 , 0.3310 , -0.0076,0.02969, -0.9130 , -0.8845]

b) [ 1.8268, 0.1200, -2.2552 , 0.3310 , -0.0076,0.02969, -0.9130 , -2.8845]

c) [ 0.8404, 0.1123 , -0.4121, 0.2159, -0.0178,-0.01480, -0.8739, -0.8391]

d) [ 0.9268, 0.3200, -0.2552 , 0.4310 , -0.0076,0.02969, -0.9130 , -1.8845]

Correct Answer: c

In [13]:
import numpy as  np
from sklearn.pipeline import Pipeline
from sklearn.linear_model import SGDRegressor
from sklearn.model_selection import train_test_split

def baseline_sgd( X, y, X_test, y_test, scaler):
  model = Pipeline([('scale',scaler),('sgd',SGDRegressor(random_state=0))])
  model.fit(X, y)

  return_vals= {"Coeffecients": model[-1].coef_,"Bias": model[-1].intercept_,
              }

  return return_vals
X_train,Xtest,y_train,ytest= train_test_split(X,y, test_size=0.2, shuffle=False, random_state=42)
baseline_sgd(X_train,y_train,X_test,y_test,StandardScaler())

{'Coeffecients': array([ 0.84046697,  0.112331  , -0.41213039,  0.21595971, -0.01781887,
        -0.01480892, -0.87394103, -0.83913104]),
 'Bias': array([2.01123921])}